source: implement OBS media status and control callbacks

kamalmostafa opened the pull request:
I needed these features for my websocket-controlled OBS application.  Looks like other folks are looking for them too!

Fixes #70

-----

Enables multiple OBS-websocket methods and OBS GUI "Source Toolbar" controls for obs-gstreamer sources.

Implements the following OBS media status callbacks:
```
    media_get_state
    media_get_time
    media_get_duration
    media_play_pause
    media_stop
    media_restart
    media_set_time (for seek-enabled pipelines)
```

fzwoch commented on gstreamer-source.c:
```diff
@@ -308,6 +308,108 @@ const char *gstreamer_source_get_name(void *type_data)
+	data_t *data = user_data;
+
+	GstState state = pause ? GST_STATE_PAUSED : GST_STATE_PLAYING;
+	gst_element_set_state(data->pipe, state);
```
Could have just used the ternary without an extra variable, but I don't mind that either.

kamalmostafa replied:
Dropped the 'state' variable in my [v2].

fzwoch commented:
Thanks, this looks quite nice already.

I could imagine to extend the state report with `OBS_MEDIA_STATE_BUFFERING` and `OBS_MEDIA_STATE_ENDED`. It would require a set and clear of a state variable, coming into the bus callback. Can quickly slip to cover all cases though. The first one would be for `GST_MESSAGE_BUFFERING` and the second one `GST_MESSAGE_EOS`. But I would accept it as it is already.

kamalmostafa commented on gstreamer-source.c:
```diff
@@ -308,6 +308,108 @@ const char *gstreamer_source_get_name(void *type_data)
+		return OBS_MEDIA_STATE_PLAYING;
+	}
+
+	return 0;
```
Agreed.  (OBS_MEDIA_STATE_NONE == 0 anyway, but this resolved in my [v2] regardless.)

kamalmostafa commented:
Thanks for your review Florian :+1: ...

[v2 notes]

I took a start at implementing a special tracking variable to deal with EOS and BUFFERING (and fussing about when to clear those), but then it occurred to me that we could just track *all* the obs_media_states  within `bus_callback()`.  So I've implemented that scheme instead and verified that it allows OBS to detect EOS and some additional error conditions as well.

I've also applied the changes you requested -- and fixed  `_restart` to not pass through GST_STATE_NULL (which works better).

I'm pushing a (rebased) stack of commits here.  The first commit (source: implement OBS media status and control callbacks) is unchanged in this stack, to ease your review of the new changes -- but my expectation is that I (or you) will squash them all down into the original commit, if you approve of the bus_callback scheme.

fzwoch commented on gstreamer-source.c:
```diff
@@ -67,11 +68,50 @@ static gboolean start_pipe(gpointer user_data)
+			break;
+		}
+		break;
+	case GST_MESSAGE_BUFFERING:
```
I think it is required to parse the actual message. If the value from the message is `!= 100` it is basically buffering, when it is `== 100` when buffering is complete and flag should be cleared.

kamalmostafa replied:
Yes, I'll take another look at this.

fzwoch commented on gstreamer-source.c:
```diff
@@ -316,6 +356,88 @@ const char *gstreamer_source_get_name(void *type_data)
+{
+	data_t *data = user_data;
+
+	gst_element_set_state(data->pipe, GST_STATE_PLAYING);
```
I'm not sure what is expected for the pipeline. Is this a NOP when it is playing or should it seek back to the start and start over?

If it is supposed to start over probably something similar has to be done as in the bus callback EOS case and the restart flag being set. E.g. stoping the pipeline and rebuild it. (Note that simply changing states seems not enough as that breaks decodebin and friends)

kamalmostafa replied:
Yes, you're right -- this only appeared to be "working" in my initial testing because of what I (now) think is another bug!...

[ other bug (?) ]   When a bus error message occurs, the fallthrough to the EOS case from the ERROR case *always* sets the state to GST_STATE_NULL immediately -- before even checking the restart_on_ flags at all.  Should it really do that?  Quick smoke testing suggests this might be why my rtmp cameras (but not my rtsp cameras) usually flail quite a bit at first startup, requiring multiple restart attempts before finally getting lucky and starting properly.

I will experiment further with "[ other bug ]".

And I'll take another look at making `_restart` actually perform a full pipeline restart.

fzwoch replied:
I think when an error occurs it is usually an indication that the pipeline cannot continue. So I _believe_ it is okay to stop.

But if you see that in your RTMP cameras it may be interesting to know what these errors are.

The rebuilding of the pipeline is usually necessary for any elements that have "sometimes pad" like decodebin. The parse_launch methods will install signal handlers for connecting these pads when they become available. When you NULL->PLAY these again these handlers are no longer effective causing the pipeline to not connect correctly.

So maybe that is what you are seeing?

kamalmostafa replied:
Ok, turns out that my rtmp-camera-flailing problem was not caused by that GST_STATE_NULL state -- but it also turns out that I would *still* like to defer that state change anyway, to allow for "rewind" even after reaching EOS.  That is implemented in the first patch of my [v3] below.

kamalmostafa replied:
[v3 notes]

This is working great now for all OBS control and state-query operations, including detection of pipeline buffering.  All OBS media controls work as they should and (corrected from [v2]) the "restart" operation now completely destroys and recreates the gstreamer pipe.
 
Fwiw, the OBS gui does not actually show the buffering state in any way, but by adding extra logging to obs-gstreamer (not included here) I can verify that OBS websocket-query state does correctly indicate OBS_MEDIA_STATE_BUFFERING as long as the gstreamer pipeline is buffering.

I'm ready for this to be merged if you are.

fzwoch commented on gstreamer-source.c:
```diff
@@ -67,11 +77,53 @@ static gboolean start_pipe(gpointer user_data)
+					    ? OBS_MEDIA_STATE_BUFFERING
+					    : OBS_MEDIA_STATE_PAUSED;
+			break;
+		case GST_STATE_PLAYING:
```
I think you need to make the same buffering logic as for PAUSED. A pipeline does not change its state just because it is buffering.

If it is desired to pause and wait, the application or code would have to do so. But that would increase latency for that stream. That is something for another day though.

fzwoch replied:
Or actually, maybe it would be more correct to just save the state here and when the actual sate is requested, first check for the `buffering` flag, return `OBS_MEDIA_STATE_BUFFERING` if it is set, and only return the real state if the flag is cleared.

kamalmostafa replied:
You're right on both counts.  I was not doing the "buffering override" for PLAYING because it seemed to cause bad behavior which I now realize was actually because of your second comment...  It is actually required that we evaluate our buffering override at query-time because we might not get any further state change messages after receiving the "buffering=100%" message.

Addressed in my [v4] push.

fzwoch commented on gstreamer-source.c:
```diff
@@ -82,7 +134,6 @@ static gboolean bus_callback(GstBus *bus, GstMessage *message,
 		g_error_free(err);
 	} // fallthrough
 	case GST_MESSAGE_EOS:
-		gst_element_set_state(data->pipe, GST_STATE_NULL);
```
I hope this is safe? Just wondering if something may continue to do something bad here. Probably not, but I have not tested :)

kamalmostafa replied:
I've tested this pretty extensively, and I'm reasonably certain that it's not required (or even useful really) to set GST_STATE_NULL "immediately" for an error condition here -- and it is very undesirable for the EOS condition.   It's fine to leave the pipeline in ERROR (forever or until it we destroy and recreate it).   It's very userful to leave the pipeline in the EOS state, because we can still seek back and resume playing -- works great in my testing with the OBS slider control (and watching with OBS-websocket to see the transition from "playing" to "eos" and then from "eos" back to "buffering->playing" when I rewind-seek).

I will go so far as to say that I believe we are required to set GST_STATE_NULL immediately before unref'ing data->pipe *and never at any other time than that*.

My [v4] implements such.

fzwoch commented on gstreamer-source.c:
```diff
@@ -316,6 +367,86 @@ const char *gstreamer_source_get_name(void *type_data)
+enum obs_media_state gstreamer_source_get_state(void *user_data)
+{
+	data_t *data = user_data;
+
```
if  mentioned above, probably:
```C
data_t *data = user_data;

if (data->buffering)
  return OBS_MEDIA_STATE_BUFFERING;

return data->obs_media_state;
```
What about STOPPED but buffer was currently done? Clear buffering flag on EOS, ERROR?

fzwoch commented on gstreamer-source.c:
```diff
@@ -316,6 +367,86 @@ const char *gstreamer_source_get_name(void *type_data)
+
+int64_t gstreamer_source_get_time(void *user_data)
+{
+	data_t *data = user_data;
```
I wonder if all of these should check again data->pipe not being NULL.

kamalmostafa replied:
Yup.  [v4] adds the missing `if (data->pipe)` guards to the media callbacks that should have them.

fzwoch commented on gstreamer-source.c:
```diff
@@ -48,16 +50,24 @@ static void timeout_destroy(gpointer user_data)
+	gst_element_set_state(data->pipe, GST_STATE_NULL);
+
 	gst_object_unref(data->pipe);
 	data->pipe = NULL;
```
I think this may be handy..
```c
data->obs_media_state = OBS_MEDIA_STATE_NONE;
data->buffering = false;
```

fzwoch commented:
I commented a lot.. but I think it is actually quite close..

kamalmostafa commented:
I'll comment on the whole "buffering" issue later -- but first -- a new general problem with my OBS media callback design:

It turns out that it is unsafe to destroy/reconstruct **directly** from the OBS media callbacks as I've been doing.  Those callbacks are being called from _some_ OBS thread -- I don't really know if its the same thread that initially started the plugin -- it might be some other GUI thread attached to the media-control buttons (?).  So once we click the OBS "Stop" and "Restart" buttons, the new bus callback we set up ends up running in the context of "that thread".  We've been getting away with it anyway, because whatever thread that is keeps running as long as OBS is running (?).  That's pretty icky right there, but it appears to function normally anyway.

BUT ... A serious problem is exposed when we use OBS-websocket to invoke our stop and restart.  The callbacks run okay, and the pipeline restarts and video plays -- but now our newly created bus callback has ended up running in the context of the **OBS-websocket** thread.   And if the application connected to OBS-websocket then chooses to disconnect from OBS, then that thread just goes poof!  Weirdly, in this state, our pipeline will *still* be playing video just fine -- but our bus callback disappears, and we lose the ability to read the media state altogether.

Upshot of all this: Our OBS media stop/restart callbacks must arrange for the teardown and reconstruct to run in the context of the main loop thread, like e.g.
```
        g_main_context_invoke(g_main_loop_get_context(data->loop),
                              start_pipe, data);
```

I have that core idea working, but I'll need to re-think about when the media states get reset and clean it up.  So [v4] will be on the way soon.

fzwoch commented:
Sounds like something I started to do for https://github.com/fzwoch/obs-gnome-screencast.

But I forgot why I actually needed it to do there. But yeah, sounds like the right thing to do, something like the `g_idle_add()` in the callback itself, just that you additionally need to select the correct context.

kamalmostafa commented:
Here's my [v4], which addresses all of your comments from last round (I think) -- and fixes all of the `obs_media_set_*` callbacks to run their work in the obs-gstreamer loop context.

It's getting to be quite a bit of code, so I'm recommend leaving it as separate commits; each commit here is a valid functionality point.

I'm not aware of any problems with this [v4], but I'll advise you whether I find any over the next couple of days running in my real-world application.

kamalmostafa commented:
FYI, The extra force-push of the `indicate "buffering"` commit there was just to amend the git comment (no code change).

fzwoch commented:
Looks good to me and sounds like reasonable tested. If we notice a bug it can always be fixed. Good work, thank you!

kamalmostafa commented:
:+1: Thanks for obs-gstreamer, Florian -- and thanks for merging this!

rodrigograca31 commented:
Just wanted to report its working great! (at least the restart 😂)(on Linux)
Thanks @kamalmostafa for the hard work!
